apps/music/js/queue.js

```diff
@@ -0,0 +1,335 @@
+
+      this._shuffledList = fillIndices(this.length);
+      var oldIndex = null;
+      if (keepIndex) {
+        oldIndex = this._shuffledList.splice(this._index, 1)[0];
+        this._index = 0;
+      }
+
+      shuffle(this._shuffledList);
+
+      if (keepIndex) {
+        this._shuffledList.unshift(oldIndex);
+      }
+    },
+
+    /**
+     * Move the queue's current pointer to the next track.
+     *
+     * @param {Boolean} automatic True if this action is happening automatically
+     *        (i.e. when a track ends).
+     */
+    next: function(automatic = false) {
+      return this.advance(1, automatic);
+    },
+
+    /**
+     * Move the queue's current pointer to the previous track.
+     */
+    previous: function() {
+      this.advance(-1);
+    },
+
+    /**
+     * Move the queue's current pointer to another track.
+     *
+     * @param {Number} offset The offset to move by.
+     * @param {Boolean} automatic True if this action is happening automatically
+     *        (i.e. when a track ends).
+     */
+    advance: function(offset, automatic = false) {
+      var next;
+      var justShuffled = false;
+
+      // If the shuffle setting was changed after we last updated, shuffle
+      // (or unshuffle, as necessary).
+      if (this._shuffleTime < shuffleTime) {
+        this._shuffle(playbackSettings.shuffle);
+        justShuffled = true;
+      }
+
+      switch (PlaybackQueue.repeat) {
+      case Repeat.OFF:
+        next = Math.max(this._index + offset, 0);
+        break;
+      case Repeat.LIST:
+        next = posmod(this._index + offset, this.length);
+        var cycled = next !== this._index + offset;
+        // If we cycled around the list, reshuffle if necessary.
+        if (playbackSettings.shuffle && cycled && !justShuffled) {
+          this._shuffle(true);
+        }
+        break;
+      case Repeat.SONG:
+        next = automatic ? this._index : Math.max(this._index + offset, 0);
+        break;
+      default:
+        throw new Error('unexpected repeat status: ' + this.repeat);
+      }
+
+      // If we hit the end, just invalidate the queue and return.
+      if (next >= this.length) {
+        this._index = null;
+        return false;
+      }
+
+      this._index = next;
+      return true;
+    },
+
+    /**
+     * Get the index of the current track.
+     */
+    get index() {
+      return this._shuffledList ? this._shuffledList[this._index] : this._index;
+    },
+
+    /**
+     * Get the raw index of the current track (i.e. don't shuffle the items).
+     */
+    get rawIndex() {
+      return this._index;
+    }
+  };
+
+  /**
+   * A playback queue for statically-defined queues.
+   *
+   * @param {Array} fileinfos An array of file infos for the tracks.
+   * @param {Number} [index] The index of the queue to start at.
+   */
+  function StaticQueue(fileinfos, index) {
+    this._fileinfos = fileinfos;
+    this._init(index);
+  }
+
+  StaticQueue.prototype = new BaseQueue();
```

Comment:
I think I'll leave this, since it seems ok for now. In the future, we should switch to using real classes, though.